whats-that-typeclass-Foldable/article.md

```diff
@@ -0,0 +1,359 @@
+
+### Strict `foldl'`
+
+Haskell uses lazy evaluations by default, that have a positive effect on performance in a lot of cases. Although, the laziness might impact negatively sometimes, and folding is this case exactly. Imagine that you want to sum up the elements in the really huge list. When using `foldl` or `foldr`, a value of an accumulator isn't evaluated on each step, so [thunk](https://wiki.haskell.org/Thunk) is being accumulated. Considering `foldl`, in the first step thunk is just ``z `f` x₁``, in the second — ``(z `f` x₁) `f` x₂``, that's not a big deal. But in the final step, an accumulator stores ``( ... (z `f` x₁) `f` x₂) `f` ...) `f` xₙ₋₁) `f` xₙ``. If the list's length is above ~10^8, thunk becomes too large to store it in a memory, and we get `** Exception: stack overflow`. However, that's not the reason to drop Haskell, as we have a rescue `foldl'`!
+
+`foldl'` provides forcing to [Weak Head Normal Form](https://wiki.haskell.org/Weak_head_normal_form) in each step, that prevents thunk from being accumulated. So `foldl'` is often a desirable way to reduce a container, especially when it's large, and you want a final strict result.
+
+Here are some "benchmarks", that may vary depending on a computer's characteristics. But they are illustrative enough to grasp the difference between the performance of strict and lazy fold.
+
+```haskell
+-- set GHCi option to show time and memory consuming
+ghci> :set +s
+
+-- lazy `foldl` and `foldr`, length = 10^7
+ghci> foldl (+) 0 [1..10000000] 
+50000005000000
+(1.97 secs, 1,612,359,432 bytes)
+
+ghci> foldr (+) 0 [1..10000000] 
+50000005000000
+(1.50 secs, 1,615,360,800 bytes)
+
+-- lazy `foldl` and `foldr`, length = 10^8
+ghci> foldl (+) 0 [1..100000000]
+*** Exception: stack overflow
+
+ghci> foldr (+) 0 [1..100000000]
+*** Exception: stack overflow
+
+-- strict `foldl'`, length = 10^8..10^10
+ghci> foldl' (+) 0 [1..100000000]
+5000000050000000
+(1.43 secs, 8,800,060,520 bytes)
+
+ghci> foldl' (+) 0 [1..1000000000]
+500000000500000000
+(15.28 secs, 88,000,061,896 bytes)
+
+ghci> foldl' (+) 0 [1..10000000000]
+50000000005000000000
+(263.51 secs, 1,139,609,364,240 bytes)
+```
+
+### Others
+
+There are other methods of `Foldable` type class that could be helpful sometimes. For example, there is a `fold` having `Monoid` constraint, combining elements of a container with `(<>)`. `foldl1` and `foldr1`, that should be used with non-empty containers only. The `length` function, which you might know, is also a member of the `Foldable`, along with `maximum`, `minimum`, and `null`. You might proceed to [the documentation](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-Foldable.html) to get to know them better.
+
+## Exercises to check yourself
+
+The theoretical part of our article is over. Hope you know what's that typeclass `Foldable` now! We suggest you to check yourself with our mini exercises.
+
+### Which one definition is correct? 
```

Comment:
It seems odd to have task definitions as subheadings, perhaps there is a nicer formatting like lists, or table?

Reply:
[I've replaced](https://github.com/serokell/blog-posts/pull/3/commits/4a1b8c8763718ce93344782165af0272c9d908a1) them with list

Reply:
Seems good!